app/scripts/modules/core/src/cluster/filter/ClusterFilters.tsx

```diff
@@ -0,0 +1,334 @@
+import React from 'react';
+import { useOnStateChanged } from '@uirouter/react';
+import { compact, uniq, map } from 'lodash';
+
+import { Application } from 'core/application';
+import { ClusterState } from 'core/state';
+import { FilterCheckbox, ISortFilter, digestDependentFilters } from 'core/filterModel';
+import { robotToHuman, useDataSource, useObservable } from 'core/presentation';
+import {
+  buildLabelsMap,
+  labelFiltersToTrueKeyObject,
+  trueKeyObjectToLabelFilters,
+  ILabelFilter,
+} from './labelFilterUtils';
+
+import { poolBuilder } from './clusterDependentFilterHelper.service';
+import { FilterSection } from './FilterSection';
+import LabelFilter from './LabelFilter';
+
+export interface IClusterFiltersProps {
+  app: Application;
+}
+
+interface IClusterHeaders {
+  account: string[];
+  availabilityZone: string[];
+  category: string[];
+  detail: string[];
+  instanceType: string[];
+  providerType: string[];
+  region: string[];
+  stack: string[];
+}
+
+export const ClusterFilters = ({ app }: IClusterFiltersProps) => {
+  const { serverGroups } = app;
+  const { data: serverGroupData, loaded: clustersLoaded } = useDataSource(serverGroups);
+
+  const [tags, setTags] = React.useState(ClusterState.filterModel.asFilterModel.tags);
+  const [sortFilter, setSortFilter] = React.useState<ISortFilter>(ClusterState.filterModel.asFilterModel.sortFilter);
+  const [labelFilters, setLabelFilters] = React.useState<ILabelFilter[]>(
+    trueKeyObjectToLabelFilters(sortFilter.labels),
+  );
+
+  const labelsMap = buildLabelsMap(serverGroupData);
+  const showLabelFilter = Object.keys(labelsMap).length > 0;
+
+  const getHeadingsForOption = (option: string): string[] =>
+    compact(uniq(map(serverGroupData, option) as string[])).sort();
+  const [headings, setHeadings] = React.useState<IClusterHeaders>({
+    account: [],
+    availabilityZone: [],
+    category: getHeadingsForOption('category'),
+    detail: ['(none)'].concat(getHeadingsForOption('detail')),
+    instanceType: [],
+    providerType: [],
+    region: [],
+    stack: ['(none)'].concat(getHeadingsForOption('stack')),
+  });
+
+  useObservable(ClusterState.filterService.groupsUpdatedStream, () => {
+    setTags(ClusterState.filterModel.asFilterModel.tags);
+    setSortFilter(ClusterState.filterModel.asFilterModel.sortFilter);
+  });
+
+  useOnStateChanged(() => {
+    ClusterState.filterModel.asFilterModel.activate();
```

Comment:
Should this be checking if the new state is actually the Clusters State?

Reply:
That is a good point. The previous version does not do that, but adding that check makes sense from an efficiency standpoint.